### ISSUE TITLE
fliplr not working as expected

### ISSUE THREAD

**Issue by MasIgor** [ultralytics/ultralytics#5332]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Bug

Hi all,

maybe i did something wrong, but here is what happened: 

I enabled the augmentation during training, and edited the /ultralytics/cfg/default.yaml file. I do have left and right objects, so i left fliplr at 0.0 and set flipud at 0.5. 

The problem is that the resulting training is unsure about left and right versions of the object.

the trainig terminated with these results as last epoch:

 Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
    120/120      15.3G    0.08112    0.08898     0.8109         45        640: 1
                 Class     Images  Instances      Box(P          R      mAP50  m
                   all       1846       1846      0.998      0.999      0.995      0.995

but the final validation gave all results like this:

 Class     Images  Instances      Box(P          R      mAP50  m
                   all       1846       1846      0.531          0.998      0.759      0.759
                 23-R       1846         26        0.5          1      0.507      0.507
                 23-L       1846         26      0.499          1      0.553      0.553
                 25-R       1846         26        0.5          1      0.539      0.539
                 25-L       1846         26      0.493      0.972      0.599      0.599
                 28-R       1846         26      0.509          1      0.581      0.581
                 28-L       1846         26        0.5          1      0.618      0.618
                 29-R       1846         26        0.5          1      0.527      0.527
                 29-L       1846         26      0.499          1      0.528      0.528

because it flipped them during training. 

so my questions are: 

1) why is there this discrepance between epochs Precision and validation Precision?
2) why is it flipping the images horizontally even if the default.yaml is set up like this:

# Hyperparameters ------------------------------------------------------------------------------------------------------
lr0: 0.01  # (float) initial learning rate (i.e. SGD=1E-2, Adam=1E-3)
lrf: 0.01  # (float) final learning rate (lr0 * lrf)
momentum: 0.937  # (float) SGD momentum/Adam beta1
weight_decay: 0.0005  # (float) optimizer weight decay 5e-4
warmup_epochs: 3.0  # (float) warmup epochs (fractions ok)
warmup_momentum: 0.8  # (float) warmup initial momentum
warmup_bias_lr: 0.1  # (float) warmup initial bias lr
box: 7.5  # (float) box loss gain
cls: 0.5  # (float) cls loss gain (scale with pixels)
dfl: 1.5  # (float) dfl loss gain
pose: 12.0  # (float) pose loss gain
kobj: 1.0  # (float) keypoint obj loss gain
label_smoothing: 0.0  # (float) label smoothing (fraction)
nbs: 64  # (int) nominal batch size
hsv_h: 0.015  # (float) image HSV-Hue augmentation (fraction)
hsv_s: 0.7  # (float) image HSV-Saturation augmentation (fraction)
hsv_v: 0.4  # (float) image HSV-Value augmentation (fraction)
degrees: 0.0  # (float) image rotation (+/- deg)
translate: 0.1  # (float) image translation (+/- fraction)
scale: 0.5  # (float) image scale (+/- gain)
shear: 0.0  # (float) image shear (+/- deg)
perspective: 0.0  # (float) image perspective (+/- fraction), range 0-0.001
flipud: 0.5  # (float) image flip up-down (probability)
fliplr: 0.0  # (float) image flip left-right (probability)
mosaic: 1.0  # (float) image mosaic (probability)
mixup: 0.0  # (float) image mixup (probability)
copy_paste: 0.0  # (float) segment copy-paste (probability)

Thank you

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by MasIgor** [comment]
**### Update.**

sorry it was my fault. I edited the default.yaml of the git clone instead of the installed one. But the other question still remains:

why is there the discrepancy between the epoch precision and the validation precision? it is as if the augmentation is only applied durng validation... 

thank you

**Comment by glenn-jocher** [comment]
@MasIgor hello,

Thank you for reaching out. The discrepancy between the epoch precision and the validation precision might be occurring because the network is seeing flipped images for the first time at the validation stage, despite the specified configuration in the hyperparameters section.

The training (epoch) precision represents how well the model is doing on your training set. However, the validation precision shows how well the model performs on unseen data (the validation set). The discrepancy could be a result of overfitting or data leakage between training and validation set.

During augmentation, the model is supposed to learn variants of the object, so at the validation/testing time, it is able to detect the object regardless of flip or other transformations. So if the augmentation is properly applied during training, your model should be invariant to such transformations during validation. 

Please ensure that your configuration changes are appropriately applied during training, and also make sure your data set does not have any unintentional biases or leakages. Remember to use a separate, untouched dataset for the validation phase to ensure an accurate evaluation of your model.

Thanks for your understanding! Let me know if you have further questions.

**Comment by MasIgor** [comment]
Hi Glenn!

thanks for your answer.. something is off here tough. the dataset is OK, i have already trained other networks with this dataset. the dataset contains left and right version of the same object and is labeled accordingly. if you flip it horizontally, then a left object becomes a right object but is labeled as left object. so if images get flipped horizontally with 50% chance, i would expect it to have 50% of the images labeled wrong and to have the above results, so a precision of 0.5. That does make sense. 

what does not make sense to me, is that the training and dataset have similar content, so the same crates (different, unseen photos)  all correctly labeled as left and right. So IF the left/right flip is applied during training, then it should be reflected in all epochs. if it is applied only on the validation, then the training is correct, but the validation is not since it is flipping images and assignign the wrong class to it. from my result i am very confident that the left/right flip is only applied to the validation... can that be? does the validation use different hyperparameters than the training?

**Comment by sarpx** [comment]
@glenn-jocher Mixing left and right is a problem I also have.

**Comment by MasIgor** [comment]
@glenn-jocher I have an update.. i tried the trained network, and it recognizes the left and right ones just right. so editing the default.yaml file seems to have worked. this confirms even more my theory that the validation at the end of the training does flip the images left and right regardless of what has been done during training.

**Comment by glenn-jocher** [comment]
@MasIgor hey there,

It seems that there might be an inconsistency between the transformations applied during the training and validation periods. Ordinarily, the model should use the same configurations and transformations for both the training and validation phases. This ensures that performance during training is consistent with validation, and your model generalizes well to different environments. If your data experiences different transformations at validation time compared to training time, it would definitely affect the model performance on the validation set.

However, since you mentioned that the trained model is recognizing the left and right objects correctly on different unseen images, it suggests that your training process worked as expected. It's possible that the issue arises from the validation process during training, where the left-right flip might be erroneously applied irrespective of the configuration settings. 

This inconsistency isn't ideal, of course, and we appreciate your patience and understanding as we work to confirm and address the issue. In the meantime, you may want to manually control the transformations during validation to avoid any unexpected behavior.

Thank you for raising this issue, as it helps us to improve the performance and robustness of the model.
